app/Http/Controllers/Admin/ModulesController.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+namespace App\Http\Controllers\Admin;
+
+use App\Contracts\Controller;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Artisan;
+use Illuminate\Support\Facades\File;
+use Illuminate\Support\Str;
+use Laracasts\Flash\Flash;
+use Madnest\Madzipper\Madzipper;
+use App\Models\Module;
+
+class ModulesController extends Controller
+{
+    /**
+     * Display a listing of the resource.
+     *
+     * @param Request $request
+     *
+     * @return mixed
+     */
+    public function index()
+    {
+        $modules = Module::all();
+        return view('admin.modulesmanager.index', [
+            'modules' => $modules
+        ]);
+    }
+
+    public function addModule(Request $request)
+    {
+        if ($request->isMethod('get')) {
+            return view('admin.modulesmanager.add');
+        }
+        elseif ($request->isMethod('post')) {
+            $this->addModulePost($request);
+        }
+        Flash::success('Module Installed Successfully!');
+        return redirect(route('admin.modulesmanager.index'));
+    }
+
+    public function addModulePost($request)
+    {
+        $zipper = new Madzipper;
+
+        $file = $request->file('module_file');
+
+        $module = Str::replaceLast('.zip', '', $file->getClientOriginalName());
+
+        $moduleFolder = $zipper->make($file)->folder($module);
```

Comment:
The better idea will be to extract the zip into a temporary folder in `storage/`, and then read the `module.json` file for the module name, and then create the folder - or detect if a folder is at the root of the extracted files (e.g, `module_name/module.json`, then just move the entire thing.